modules/core/src/main/java/org/apache/ignite/internal/managers/encryption/GridEncryptionManager.java

```diff
@@ -734,18 +884,309 @@ private void writeToMetaStore(int grpId, byte[] encGrpKey) {
+
+        try {
+            getSpi().setMasterKeyName(rec.getMasterKeyName());
+
+            for (Map.Entry<Integer, byte[]> entry : rec.getGrpKeys().entrySet())
+                grpEncKeys.computeIfAbsent(entry.getKey(), k -> getSpi().decryptKey(entry.getValue()));
+
+            forceWriteAllKeysToMetaStore = true;
+        } catch (IgniteSpiException e) {
+            log.warning("Unable to apply group keys from WAL record [masterKeyName=" + rec.getMasterKeyName() + ']', e);
+        }
+    }
+
+    /**
+     * @param masterKeyName Master key name.
+     * @return Master key digest.
+     * @throws IgniteException if unable to get master key digest.
+     */
+    private byte[] masterKeyDigest(String masterKeyName) {
+        byte[] digest;
+
+        masterKeyChangeLock.writeLock().lock();
+
+        try {
+            String curName = getSpi().getMasterKeyName();
+
+            try {
+                getSpi().setMasterKeyName(masterKeyName);
+
+                digest = getSpi().masterKeyDigest();
+            } catch (Exception e) {
+                throw new IgniteException("Unable to set master key locally [masterKeyName=" + masterKeyName + ']', e);
+            } finally {
+                getSpi().setMasterKeyName(curName);
+            }
+        }
+        finally {
+            masterKeyChangeLock.writeLock().unlock();
+        }
+
+        return digest;
+    }
+
+    /**
+     * @param keyName Master key name to encrypt.
+     * @return Encrypted master key name.
+     */
+    private byte[] encryptKeyName(String keyName) {
+        return withMasterKeyChangeReadLock(() -> {
+            Serializable key = getSpi().create();
+
+            byte[] encKey = getSpi().encryptKey(key);
+
+            byte[] serKeyName = U.toBytes(keyName);
+
+            ByteBuffer res = ByteBuffer.allocate(/*Encrypted key length*/4 + encKey.length +
+                getSpi().encryptedSize(serKeyName.length));
+
+            res.putInt(encKey.length);
+            res.put(encKey);
+
+            getSpi().encrypt(ByteBuffer.wrap(serKeyName), key, res);
+
+            return res.array();
+        });
+    }
+
+    /**
+     * @param data Byte array with encrypted a master key name.
+     * @return Decrypted master key name.
+     */
+    private String decryptKeyName(byte[] data) {
+        return withMasterKeyChangeReadLock(() -> {
+            ByteBuffer buf = ByteBuffer.wrap(data);
+
+            int keyLen = buf.getInt();
+
+            byte[] encKey = new byte[keyLen];
+
+            buf.get(encKey);
+
+            byte[] encKeyName = new byte[buf.remaining()];
+
+            buf.get(encKeyName);
+
+            byte[] serKeyName = getSpi().decrypt(encKeyName, getSpi().decryptKey(encKey));
+
+            return U.fromBytes(serKeyName);
+        });
+    }
+
+    /** Checks that the master key change process supported by all nodes in cluster. */
+    public void checkMasterKeyChangeSupported() {
```

Comment:
This function has only one usage. Let's inline it.

Reply:
Done.